### DISCUSSION TITLE
Extensibility framework

### DISCUSSION THREAD

**Discussion by hlcianfagna** [crate/crate#13444] [Ideas]
It used to be possible to create plugins for CrateDB
https://github.com/crate/crate-example-plugin

Since UDF support was introduced the need for such an extensibility framework is much reduced, but since UDFs cannot currently access (for understandable reasons) the database, external resources, or leverage compute resources across the cluster, there may be cases where a plugins interface would still be useful. This may be for code specific to a certain application, or code that may not be of a good enough quality to distribute with the base build.

**Comment by mfussenegger** [comment]
> It used to be possible to create plugins for CrateDB

At least in theory it's still possible. E.g. the COPY FROM/TO s3 functionality is currently still a plugin, except that it's bundled

>  there may be cases where a plugins interface would still be useful.

Do you have more concrete examples?

> but since UDFs cannot currently access (for understandable reasons) the database, external resources, or leverage compute resources across the cluster

Part of the reason for these restrictions is that any blocking IO would severly impact performance. Scalar functions should be primarily CPU bound, otherwise the engine is going to handle them badly as they can be used in all sorts of places where blocking IO is trouble.
